### PR TITLE
Ports /tg/ spatial grid system, refactors telecomms, destroys lag

### PR DESCRIPTION
**Pull request by JohnWildkins**
Ports tgstation/tgstation#33647
Ports tgstation/tgstation#61422
Ports Baystation12/Baystation12#29220

Spatial grid system made by people smarter than me destroys procs like get_mobs_or_objs_in_view that eats all of our server performance on highpop

This should (hopefully) result in a smoother experience especially with high amounts of players (read: events)

Benchmarking:

~3.4-4x faster than get_mobs_or_objs_in_view on Runtime
~2.5x-10x faster than get_mobs_or_objs_in_view on Horizon (depending on range)
~3x faster to get hearers in radio ranges

todo: 
 - actually test it lmao

### CODE REVIEW

**Comment by JohnWildkins** [comment]
More or less feature complete at this point, but entirely possible it's horrifically broken in several places since I refactored it about half a dozen times.

**Comment by JohnWildkins** [comment]
!review

File: maps/away/ships/sol_pirate/sfa_patrol_ship.dm [hunk 1]
```diff
@@ -7,6 +7,7 @@
 	spawn_cost = 1
 	id = "sfa_patrol_ship"
 	shuttles_to_initialise = list(/datum/shuttle/autodock/overmap/sfa_shuttle)
+	template_flags = TEMPLATE_FLAG_SPAWN_GUARANTEED
```
Comment (NonQueueingMatt) [inline maps/away/ships/sol_pirate/sfa_patrol_ship.dm:10]: remove

**Review by NonQueueingMatt — CHANGES_REQUESTED** [review]

File: maps/away/ships/sol_pirate/sfa_patrol_ship.dm (hunk 1 above)
Comment (JohnWildkins) [reply maps/away/ships/sol_pirate/sfa_patrol_ship.dm:10]: tytyty

File: code/game/machinery/telecomms/computers/telemonitor.dm [hunk 2]
```diff
@@ -0,0 +1,161 @@
+/*
+	Telecomms monitor tracks the overall trafficing of a telecommunications network
+	and displays a hierarchy of linked machines.
+*/
+
+
+/obj/machinery/computer/telecomms/monitor
+	name = "Telecommunications Monitor"
```
Comment (SleepyGemmy) [inline code/game/machinery/telecomms/computers/telemonitor.dm:8]:
```suggestion
	name = "telecommunications monitor"
```

File: code/game/machinery/telecomms/computers/traffic_control.dm [hunk 3]
```diff
@@ -0,0 +1,238 @@
+//This file was auto-corrected by findeclaration.exe on 25.5.2012 20:42:32
+
+/obj/machinery/computer/telecomms/traffic
+	name = "Telecommunications Traffic Control"
```
Comment (SleepyGemmy) [inline code/game/machinery/telecomms/computers/traffic_control.dm:4]:
```suggestion
	name = "telecommunications traffic control"
```

File: code/game/machinery/telecomms/presets.dm [hunk 4]
```diff
@@ -136,13 +183,37 @@
...
+/obj/machinery/telecomms/server/preset_map/LateInitialize()
+	if(current_map.use_overmap && !linked)
+		var/my_sector = map_sectors["[z]"]
+		if (istype(my_sector, /obj/effect/overmap/visitable))
+			attempt_hook_up(my_sector)
+
+	if(istype(linked) && linked.comms_support)
+		var/preset_name = linked.comms_name
+		var/name_lower = replacetext(lowertext(preset_name), " ", "_")
+		id = "[preset_name] Server"
```
Comment (SleepyGemmy) [inline code/game/machinery/telecomms/presets.dm:202]:
```suggestion
		id = "[preset_name] server"
```
probably?

File: code/game/machinery/telecomms/presets.dm [hunk 5]
```diff
@@ -209,6 +280,23 @@
...
+/obj/machinery/telecomms/broadcaster/preset_map/LateInitialize()
+	if(current_map.use_overmap && !linked)
+		var/my_sector = map_sectors["[z]"]
+		if (istype(my_sector, /obj/effect/overmap/visitable))
+			attempt_hook_up(my_sector)
+
+	if(istype(linked) && linked.comms_support)
+		var/preset_name = linked.comms_name
+		var/name_lower = replacetext(lowertext(preset_name), " ", "_")
+		id = "[preset_name] Broadcaster"
```
Comment (SleepyGemmy) [inline code/game/machinery/telecomms/presets.dm:292]:
```suggestion
		id = "[preset_name] broadcaster"
```
probably?

File: code/modules/mining/mine_turfs.dm [hunk 6]
```diff
@@ -71,6 +71,9 @@ var/list/mineral_can_smooth_with = list(
 
 	turfs += src
 
+	if (isStationLevel(z))
```
Comment (SleepyGemmy) [inline code/modules/mining/mine_turfs.dm:74]:
```suggestion
	if(isStationLevel(z))
```

File: code/modules/mining/mine_turfs.dm [hunk 7]
```diff
@@ -197,6 +200,9 @@ var/list/mineral_can_smooth_with = list(
...
```
Comment (SleepyGemmy) [inline code/modules/mining/mine_turfs.dm:203]:
```suggestion
	if(isStationLevel(z))
```

File: code/modules/mining/mine_turfs.dm [hunk 8]
```diff
@@ -638,6 +644,9 @@ var/list/asteroid_floor_smooth = list(
...
```
Comment (SleepyGemmy) [inline code/modules/mining/mine_turfs.dm:647]:
```suggestion
	if(isStationLevel(z))
```

File: code/modules/mob/hear_say.dm [hunk 9]
```diff
@@ -165,15 +165,10 @@
...
 		speaker_name = "Unknown"
 
-	if(istype(speaker, /mob/living/carbon/human))
-		var/mob/living/carbon/human/H = speaker
-		if(H.voice)
-			speaker_name = H.voice
-
-	if(vname)
-		speaker_name = vname
+	if (ishuman(speaker) && speaker.GetVoice() != real_name)
```
Comment (SleepyGemmy) [inline code/modules/mob/hear_say.dm:168]:
```suggestion
	if(ishuman(speaker) && speaker.GetVoice() != real_name)
```

File: code/modules/mob/hear_say.dm [hunk 10]
```diff
@@ -165,15 +165,10 @@
...
+		speaker_name = speaker.GetVoice()
 
-	if(hard_to_hear)
+	if (hard_to_hear)
```
Comment (SleepyGemmy) [inline code/modules/mob/hear_say.dm:171]:
```suggestion
	if(hard_to_hear)
```

**Review by NonQueueingMatt — APPROVED** [review]

**Pull request merged**